File: src/main/org/nlogo/swing/WindowResizer.java
package org.nlogo.swing ;

public strictfp class WindowResizer
	extends javax.swing.JPanel
	implements
		java.awt.event.MouseListener ,
		java.awt.event.MouseMotionListener
{

	private final javax.swing.JWindow window ;
	private java.awt.Point mousePressAbsLoc ;
	private java.awt.Dimension sizeWhenPressed ;
	
	public WindowResizer( javax.swing.JWindow window )
	{
		this.window = window ;
		addMouseListener( this ) ;
		addMouseMotionListener( this ) ;
	}

	public void mouseMoved(java.awt.event.MouseEvent e ) { /* ignore */ }
	public void mouseClicked( java.awt.event.MouseEvent e ) { /* ignore */ }
	public void mouseEntered( java.awt.event.MouseEvent e ) { /* ignore */ }
	public void mouseExited( java.awt.event.MouseEvent e ) { /* ignore */ }
	public void mouseReleased( java.awt.event.MouseEvent e ) { /* ignore */ }

	public void mousePressed( java.awt.event.MouseEvent e ) 
	{
		java.awt.Point mousePressLoc = e.getPoint() ;
		mousePressAbsLoc = new java.awt.Point( mousePressLoc ) ;
		org.nlogo.awt.Utils.convertPointToScreen( mousePressAbsLoc , WindowResizer.this ) ;
		sizeWhenPressed = window.getSize() ;
	}

	public void mouseDragged( java.awt.event.MouseEvent e )
	{
		java.awt.Point dragAbsLoc = new java.awt.Point( e.getPoint() ) ;
		org.nlogo.awt.Utils.convertPointToScreen( dragAbsLoc , WindowResizer.this ) ;
		window.setSize
			( sizeWhenPressed.width + ( dragAbsLoc.x - mousePressAbsLoc.x ) ,
			  sizeWhenPressed.height + ( dragAbsLoc.y - mousePressAbsLoc.y ) ) ;
	}
}

File: src/main/org/nlogo/swing/WindowResizer.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.swing

import java.awt.{ Dimension, Point }
import javax.swing.JWindow
import java.awt.event.{ MouseAdapter, MouseMotionAdapter, MouseEvent }
import org.nlogo.awt.Coordinates.convertPointToScreen

// In Java 6 you can just use a single MouseAdapter for everything, but for Java 5 you need a
// separate MouseMotionAdapter. - ST 5/27/11

class WindowResizer(window: JWindow)
extends javax.swing.JPanel {

  private var mousePressAbsLoc: Point = null
  private var sizeWhenPressed: Dimension = null

  private val adapter = new MouseAdapter {
    override def mousePressed(e: MouseEvent) {
      val mousePressLoc = e.getPoint
      mousePressAbsLoc = convertPointToScreen(mousePressLoc, WindowResizer.this)
      sizeWhenPressed = window.getSize
    }
  }
  private val motionAdapter = new MouseMotionAdapter {
    override def mouseDragged(e: MouseEvent) {
      val dragAbsLoc =
        convertPointToScreen(e.getPoint, WindowResizer.this)
      window.setSize(
        sizeWhenPressed.width + (dragAbsLoc.x - mousePressAbsLoc.x),
        sizeWhenPressed.height + (dragAbsLoc.y - mousePressAbsLoc.y))
    }
  }

  addMouseListener(adapter)
  addMouseMotionListener(motionAdapter)

}
